### DISCUSSION TITLE
Non-Max-Suppression & Duplicate Detections in YOLO V8 Trained on Custom Data

### DISCUSSION THREAD

**Discussion by zburq** [ultralytics/ultralytics#10098] [Q&A]
I have trained YOLOv8 to detect and track some custom object classes in video data. The model works well but has some occasional quirks: 

Sometimes for a single object in an a single video frame, the model returns two detections and bounding boxes. The boxes have a large IoU %age (> 80%), different class labels with different confidences, and different track IDs. This happens quite sporadically. The end result is over-counting the detected objects.

This looks like a failure of non-max-suppression step. 

I also noticed that for the`export` method, the `nms` config is set to `False` by default:

https://docs.ultralytics.com/modes/export/#arguments

Is this also the case for the `train`, `detect` or `track` methods? Do I need to set `nms = True` in training?

Should `nms = True` be the default?

**Comment by glenn-jocher** [comment]
Hello there! 👋

Indeed, it seems like the situation you're encountering is tied to the model's non-max-suppression (NMS) behavior, particularly since you're seeing high-overlap detections with different labels and track IDs.

In YOLOv8's training (`train`), detection (`predict`), and tracking (`track`) modes, NMS is implicitly applied and doesn't require manual activation. Consequently, you won't need to set `nms = True` during training or inference. The `nms` argument you found in the `export` documentation specifically pertains to model export settings for formats like CoreML, which is separate from the detection/tracking process.

That said, occasional overlaps with different labels might suggest a need for further hyperparameter tuning (such as adjusting the confidence threshold or IoU thresholds for NMS) or additional training data to help the model better distinguish between classes.

If the issue persists, consider reviewing the overlap between your classes' characteristics in the training data. This might help reduce the sporadic dual detections by improving class distinction.

Hope this helps! If you have more questions or need further assistance, feel free to ask. Happy detecting! 😊

**Comment by zburq** [comment]
@GlennJochen,

This seems like some blind trial & error.

The IoU overlap in the examples is already pretty high - over 90% - the two
bounding boxes are almost identical. And the confidence levels of the two
classifications are close, but not identical.

I'm still getting some duplicate detections.

It's almost like the model acquires an extra track for a vehicle that is
already been tracked.
The vehicle appears in frame 0 to 10. In frame 4 or 5, a second bounding
box appears on top of the existing b'box - almost identical, with a
different class label. This second box persists till the end of the frame.
This produces an over-count of this vehicle type.

I'd like to figure out concretely what is going on - rather than take blind
stabs in the dark.

Z

**Reply by glenn-jocher** [reply]
Hey Z,

Thanks for reaching out with more details. 😊 It sounds a bit tricky, doesn't it? Since the overlaps and track IDs are showing behaviors not typically expected, especially with nearly matching bounding boxes and yet, different class labels, it tells us there might be something more at play here.

To dive deeper and reduce these duplicate detections, particularly in cases where a new track ID pops up midway through an already established tracking – consider fine-tuning the tracking parameters. For scenarios like yours, focusing on the tracking algorithm's `iou_threshold` could potentially help minimize the issue by tweaking how strictly it considers two boxes to overlap as the same track.

A suggestion would be to experiment specifically with the tracking thresholds, perhaps like this (if you're using the Python API):
```python
model.track(source, iou_threshold=0.95)
```

This is, of course, assuming that the model's behavior in distinguishing between very similar class labels might also be contributing to the problem. Do the classes that are getting confused share a lot of common features?

Your point about wanting a more concrete understanding rather than guessing is absolutely valid. If adjusting the tracking parameters doesn't clear things up, it might be beneficial to review the training data for those specific classes again and see if there's an opportunity for better class separation or if there's anything unusual that might be causing the classifier to waver.

I hope this gives a bit more direction, and I'm here if you need more detailed assistance! Keep up the good detective work – it's how we get to the bottom of these mysteries. 🕵️‍♂️
